File: src/main/java/com/revbingo/spiff/events/DebugEventListener.java
/*******************************************************************************
 * Copyright 2012 Mark Piper
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package com.revbingo.spiff.events;

import com.revbingo.spiff.datatypes.Datatype;

public class DebugEventListener implements EventListener {

	int tabCount = 0;

	private EventListener wrappedListener;
	
	public DebugEventListener() {
		this(new EventListener() {

			@Override
			public void notifyData(Datatype ins) {
			}
			
			@Override
			public void notifyGroup(String groupName, boolean start) {
			}
		});
	}
	
	public DebugEventListener(EventListener wrappedListener) {
		this.wrappedListener = wrappedListener;
	}
	
	@Override
	public void notifyData(Datatype ins) {
		for(int i=0;i<tabCount;i++) {
			System.out.print("\t");
		}
		System.out.println("[" + ins.getName() + "] " + ins.getValue());
		wrappedListener.notifyData(ins);
	}

	@Override
	public void notifyGroup(String groupName, boolean start) {
		if(!start) {
			tabCount--;
		}

		for(int i=0;i<tabCount;i++) {
			System.out.print("\t");
		}
		System.out.println((start ? ">> " : "<< ") + "[" + groupName + "]");

		if(start) {
			tabCount++;
		}
		
		wrappedListener.notifyGroup(groupName, start);
	}

}

File: src/main/java/com/revbingo/spiff/events/DebugEventListener.kt
/*
 * Copyright Mark Piper 2016
 *
 * This file is part of SPIFF.
 *
 * SPIFF is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SPIFF is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SPIFF.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.revbingo.spiff.events

import com.revbingo.spiff.datatypes.Datatype

public class DebugEventListener(val wrappedListener: EventListener): EventListener {

    var tabCount = 0

    constructor(): this(object: EventListener {
        override fun notifyData(ins: Datatype?) { }
        override fun notifyGroup(groupName: String?, start: Boolean) {}
    })

    override fun notifyData(ins: Datatype) {
        repeat(tabCount, { print("\t") })
        println("[${ins.name}] ${ins.value}")
        wrappedListener.notifyData(ins)
    }

    override fun notifyGroup(groupName: String, start: Boolean) {
        if(!start) {
            tabCount--
        }

        repeat(tabCount, { print("\t") })
        val leader = if(start) ">>" else "<<"
        println("${leader} [$groupName]")

        if(start) tabCount++

        wrappedListener.notifyGroup(groupName, start)
    }
}